vespalib/src/tests/datastore/fixed_size_hash_map/fixed_size_hash_map_test.cpp

```diff
@@ -0,0 +1,312 @@
+// Copyright Verizon Media. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+
+#include <vespa/vespalib/datastore/fixed_size_hash_map.h>
+#include <vespa/vespalib/datastore/unique_store_allocator.h>
+#include <vespa/vespalib/datastore/unique_store_comparator.h>
+#include <vespa/vespalib/stllike/hash_map.h>
+#include <vespa/vespalib/util/lambdatask.h>
+#include <vespa/vespalib/util/rand48.h>
+#include <vespa/vespalib/util/size_literals.h>
+#include <vespa/vespalib/util/threadstackexecutor.h>
+#include <vespa/vespalib/gtest/gtest.h>
+
+#include <vespa/vespalib/datastore/unique_store_allocator.hpp>
+#include <vespa/vespalib/stllike/hash_map.hpp>
+
+#include <vespa/log/log.h>
+LOG_SETUP("vespalib_fixed_size_hash_map_test");
+
+using vespalib::datastore::EntryRef;
+using RefT = vespalib::datastore::EntryRefT<22>;
+using MyAllocator = vespalib::datastore::UniqueStoreAllocator<uint32_t, RefT>;
+using MyDataStore = vespalib::datastore::DataStoreT<RefT>;
+using MyCompare = vespalib::datastore::UniqueStoreComparator<uint32_t, RefT>;
+using GenerationHandler = vespalib::GenerationHandler;
+using vespalib::makeLambdaTask;
+using vespalib::GenerationHolder;
+using vespalib::datastore::FixedSizeHashMap;
+
+class FixedSizeHashMapHeld : public vespalib::GenerationHeldBase
+{
+    std::unique_ptr<const FixedSizeHashMap> _data;
+public:
+    FixedSizeHashMapHeld(size_t size, std::unique_ptr<const FixedSizeHashMap> data);
+    ~FixedSizeHashMapHeld();
+};
+
+FixedSizeHashMapHeld::FixedSizeHashMapHeld(size_t size, std::unique_ptr<const FixedSizeHashMap> data)
+    : GenerationHeldBase(size),
+      _data(std::move(data))
+{
+}
+
+FixedSizeHashMapHeld::~FixedSizeHashMapHeld() = default;
+
+struct DataStoreFixedSizeHashTest : public ::testing::Test
+{
+    GenerationHandler _generation_handler;
+    GenerationHolder  _generation_holder;
+    MyAllocator       _allocator;
+    MyDataStore&      _store;
+    std::unique_ptr<const vespalib::datastore::EntryComparator> _comp;
+    std::atomic<FixedSizeHashMap *>  _hash_map;
+    vespalib::ThreadStackExecutor _writer; // 1 write thread
+    vespalib::ThreadStackExecutor _readers; // multiple reader threads
+    vespalib::Rand48 _rnd;
+    uint32_t _keyLimit;
+    std::atomic<long> _read_seed;
+    std::atomic<long> _done_write_work;
+    std::atomic<long> _done_read_work;
+    std::atomic<long> _found_count;
+    std::atomic<int> _stop_read;
+    size_t _modulo_limit;
+    bool _report_work;
+
+    DataStoreFixedSizeHashTest();
+    ~DataStoreFixedSizeHashTest();
+    void commit();
+    void grow();
```

Comment:
Consider writing tests that doesn't need growing. Let's keep these tests as simple as possible.

Reply:
Growing removed from test.